File: tempestweb/native/sync.py
"""Native sync capability — read+write offline sync driven from Python.

Exposes the offline read-sync stack (``client/offline/{pull,sync-status,
sw-bridge}.js``) to a ``view()``: :func:`configure` registers a named source (an
endpoint + a local table), :func:`now` runs a single-flight sync (replay the
shared write queue, then pull remote changes), :func:`status` reads the current
state and :func:`watch` streams it — mirroring how ``native.network`` exposes
connectivity. The browser side lives in ``client/native/sync.js``.

The endpoint follows the convention ``GET <url>?since=<watermark>&cursor=<cursor>``
returning ``{rows, next_cursor, server_time}``; rows merge last-write-wins over
the owner-scoped store (tombstones delete, a locally-pending newer edit wins).
"""

from __future__ import annotations

from collections.abc import AsyncIterator

from pydantic import BaseModel, ConfigDict

from tempestweb.native.dispatch import native_events, send_native_call

__all__ = [
    "SyncState",
    "SyncSummary",
    "configure",
    "now",
    "status",
    "watch",
]


class SyncSummary(BaseModel):
    """The outcome of one sync run.

    Attributes:
        sent: Mutations accepted and removed from the write queue.
        remaining: Mutations still pending upload.
        failed: Mutations dead-lettered this run.
        conflicts: Mutations moved to the conflict lane this run.
        applied: Remote rows applied by the pull.
    """

    model_config = ConfigDict(frozen=True)

    sent: int = 0
    remaining: int = 0
    failed: int = 0
    conflicts: int = 0
    applied: int = 0


class SyncState(BaseModel):
    """The observable state of a sync source.

    Attributes:
        phase: ``"idle"`` | ``"syncing"`` | ``"error"``.
        online: Last known connectivity.
        pending: Pending (unpushed) mutation count.
        last_synced_at: Epoch ms of the last successful sync, or None.
        last_summary: The last run's :class:`SyncSummary`, or None.
        error: The last error message, or None.
    """

    model_config = ConfigDict(frozen=True)

    phase: str = "idle"
    online: bool = True
    pending: int = 0
    last_synced_at: int | None = None
    last_summary: SyncSummary | None = None
    error: str | None = None


async def configure(
    name: str,
    url: str,
    database: str,
    table: str,
    *,
    key_path: str | None = None,
    owner_field: str | None = None,
    watermark_key: str | None = None,
) -> None:
    """Configure (or replace) a named sync source.

    Args:
        name: A stable name used to run/observe this source.
        url: The delta-sync endpoint (``GET`` with ``since``/``cursor`` query).
        database: The IndexedDB database name for the local table.
        table: The object-store (table) name rows merge into.
        key_path: The primary key path (default ``"id"``).
        owner_field: The owner-scoping field (default ``"owner"``).
        watermark_key: The localStorage key for the cursor (default
            ``f"{name}:watermark"``).

    Raises:
        BrowserUnavailableError: If called with no native bridge installed.
    """
    args: dict[str, str] = {
        "name": name,
        "url": url,
        "database": database,
        "table": table,
    }
    if key_path is not None:
        args["key_path"] = key_path
    if owner_field is not None:
        args["owner_field"] = owner_field
    if watermark_key is not None:
        args["watermark_key"] = watermark_key
    await send_native_call("sync.configure", args)


async def now(name: str) -> SyncSummary:
    """Run one sync for a source now (replay the queue, then pull).

    Args:
        name: The configured source name.

    Returns:
        The :class:`SyncSummary` for the run.

    Raises:
        BrowserUnavailableError: If called with no native bridge installed.
    """
    value = await send_native_call("sync.now", {"name": name})
    return SyncSummary.model_validate(value)


async def status(name: str) -> SyncState:
    """Read a source's current sync state.

    Args:
        name: The configured source name.

    Returns:
        The current :class:`SyncState`.

    Raises:
        BrowserUnavailableError: If called with no native bridge installed.
    """
    value = await send_native_call("sync.status", {"name": name})
    return SyncState.model_validate(value)


async def watch(name: str) -> AsyncIterator[SyncState]:
    """Stream a source's sync state on every change (event channel / T-EV).

    Yields the current state immediately, then a fresh :class:`SyncState` on each
    transition, until the ``async for`` loop is exited (closing the subscription).

    Args:
        name: The configured source name.

    Yields:
        Each updated :class:`SyncState`.

    Raises:
        NativeError: If the browser reports the subscription failed.
        BrowserUnavailableError: If no bridge is installed, or it does not support
            the event channel.
    """
    async for value in native_events("sync.watch", {"name": name}):
        yield SyncState.model_validate(value)

File: client/native/sync.js
// native/sync.js — configurable read+write sync capability (P2).
//
// Exposes the offline read-sync stack (client/offline/{pull,sync-status,
// sw-bridge}.js) to Python as a native capability, so a `view()` can drive and
// observe sync the same way it reads connectivity via native.network. An app
// first `configure`s a named source (an endpoint + a local table); then `now`
// runs a single-flight sync (replay the shared write queue, then pull remote
// changes), `status` reads the observable state and `watch` streams it.
//
// The endpoint follows a convention (adopted from the tempest-react-sdk sync
// engine): GET <url>?since=<watermark>&cursor=<cursor> returns
// { rows, next_cursor, server_time }. Rows merge last-write-wins over the
// owner-scoped OfflineStore (tombstones delete; a locally-pending newer edit is
// kept). Apps needing other shapes use the JS libs directly.
//
// Configuring a source also wires the service-worker bridge once, so an
// OFFLINE_PULL posted by the worker after a background drain reconciles every
// configured source with the token the page holds.

import { CapabilityError } from "./index.js";
import { httpRequest } from "./http.js";
import { getOfflineQueue } from "./offline.js";
import { createOfflineStore } from "../offline/store.js";
import { createPull, createWatermark, mergeRemoteInto } from "../offline/pull.js";
import { createSyncController } from "../offline/sync-status.js";
import { installSyncBridge, SW_MESSAGES } from "../offline/sw-bridge.js";

/** @type {Map<string, ReturnType<typeof createSyncController>>} */
const _sources = new Map();
let _bridgeWired = false;

/**
 * Build a sync controller for a configured source.
 *
 * @param {string} name   The source name (used for the default watermark key).
 * @param {Object} cfg    Source config (url, database, table, key_path, ...).
 * @param {import("./index.js").NativeDeps} [deps]   Injected deps.
 * @returns {ReturnType<typeof createSyncController>} The controller.
 */
function buildSource(name, cfg, deps) {
  const store = createOfflineStore({
    databaseName: cfg.database,
    tableName: cfg.table,
    keyPath: cfg.key_path || "id",
    ownerField: cfg.owner_field || "owner",
    indexedDB: deps && /** @type {any} */ (deps).indexedDB,
  });
  const watermark = createWatermark(cfg.watermark_key || `${name}:watermark`);
  const pull = createPull({
    pullPage: async (since, cursor) => {
      const params = [];
      if (since) params.push(`since=${encodeURIComponent(since)}`);
      if (cursor) params.push(`cursor=${encodeURIComponent(cursor)}`);
      const sep = cfg.url.includes("?") ? "&" : "?";
      const url = params.length ? `${cfg.url}${sep}${params.join("&")}` : cfg.url;
      const res = await httpRequest({ method: "GET", url, json: null, headers: {} }, deps);
      const body = (res && res.json_body) || {};
      return {
        rows: body.rows || [],
        nextCursor: body.next_cursor ?? null,
        serverTime: body.server_time ?? null,
      };
    },
    applyRemote: mergeRemoteInto(store),
    watermark,
  });
  return createSyncController({ queue: getOfflineQueue(deps), pull });
}

/**
 * Wire the service-worker bridge once so background drains trigger a pull.
 *
 * OFFLINE_PULL / REPLAY_OFFLINE_QUEUE run every source's syncNow;
 * OFFLINE_QUEUE_DRAINED just refreshes the pending counts. A no-op teardown when
 * service workers are unavailable, so this is safe to call outside a browser.
 * @returns {void}
 */
function ensureBridge() {
  if (_bridgeWired) return;
  _bridgeWired = true;
  const syncAll = () => {
    for (const controller of _sources.values()) void controller.syncNow();
  };
  installSyncBridge({
    [SW_MESSAGES.PULL]: syncAll,
    [SW_MESSAGES.REPLAY]: syncAll,
    [SW_MESSAGES.DRAINED]: () => {
      for (const controller of _sources.values()) void controller.refreshPending();
    },
  });
}

/**
 * Serialize the observable sync state into the snake_case wire shape Python
 * validates (the store uses camelCase internally).
 * @param {Object} state
 * @returns {Object}
 */
function toWireState(state) {
  return {
    phase: state.phase,
    online: state.online,
    pending: state.pending,
    last_synced_at: state.lastSyncedAt,
    last_summary: state.lastSummary,
    error: state.error,
  };
}

/**
 * Resolve a configured controller by name.
 * @param {string} name
 * @returns {ReturnType<typeof createSyncController>}
 * @throws {CapabilityError} unavailable when the source was not configured.
 */
function getController(name) {
  const controller = _sources.get(name);
  if (!controller) {
    throw new CapabilityError("unavailable", `sync source "${name}" is not configured`);
  }
  return controller;
}

/**
 * Configure (or replace) a named sync source.
 * @param {{name:string, url:string, database:string, table:string, key_path?:string, owner_field?:string, watermark_key?:string}} args
 * @param {import("./index.js").NativeDeps} deps
 * @returns {Promise<{configured:boolean, name:string}>}
 */
export async function syncConfigure(args, deps) {
  const controller =
    (deps && /** @type {any} */ (deps).syncController) ||
    buildSource(args.name, args, deps);
  _sources.set(args.name, controller);
  ensureBridge();
  return { configured: true, name: args.name };
}

/**
 * Run one sync for a source now (replay the queue, then pull).
 * @param {{name:string}} args
 * @param {import("./index.js").NativeDeps} deps
 * @returns {Promise<Object>} The run summary.
 */
export async function syncNow(args, deps) {
  const summary = await getController(args.name).syncNow();
  return summary || { sent: 0, remaining: 0, failed: 0, conflicts: 0, applied: 0 };
}

/**
 * Read a source's current sync state.
 * @param {{name:string}} args
 * @param {import("./index.js").NativeDeps} deps
 * @returns {Promise<Object>} The sync state.
 */
export async function syncStatus(args, deps) {
  return toWireState(getController(args.name).status.get());
}

/**
 * Stream a source's sync state on every change (T-EV). Emits the current state
 * immediately, then on each transition.
 * @param {{name:string}} args
 * @param {(payload:Object) => void} emit
 * @param {import("./index.js").NativeDeps} deps
 * @returns {() => void} Teardown that removes the subscription.
 */
export function syncWatch(args, emit, deps) {
  return getController(args.name).status.subscribe((state) =>
    emit({ event: toWireState(state) }),
  );
}
